Fixes

dougnazar opened the pull request:
Here are several fixes for `nrpe` & `check_nrpe`. Let me know if you'd prefer me to break this up into multiple requests. Kind of kept finding new stuff as I went along.

- Fix various warnings when building with `-Wall`.

  Various misc fixes, mainly left over code.
- Split `my_system()` into separate functions for child & parent.
- Rewrite `my_system_*()` IO loops to handle async & large buffers.

  Due to scheduling and various implementations, we can't depend on read/write having data/space available. It's also possible to fill the send buffer before we exit so need to read data while waiting for child to exit.
  I've tested with 10000, 60000, 70000 sized outputs. The 10000 & 60000 hashes match, and as expected the 70000 gets truncated.
  Will probably help with #271
- Fix debug display of acl IP addresses

  Would always show same address due to shared buffer.
- Show correct address family when setting up listen addresses.

  Would always say `IPv4` no matter which family the address was.
- Only make requested binary

  Top level Makefile didn't pass command to child so would always `make all`
- Don't compile utils.c twice, once for each program.

  `utils.c` is shared between both programs.
- Fix a few pedantic warnings, mostly prototype related.
- Fix use of uninitialized variable.
- Clear any existing ACLs before parsing new config.

  Would get duplicate ACL entries if allowed_hosts was used multiple times in config or on every `SIGHUP`.
  Also would try to dump IPv6 ACLs as IPv4.
- Fix various leaks detected by valgrind.

  All config options that use `strdup()` would leak memory if used more than once or on `SIGHUP`.
  Only allocate `fdset` when size changes and ensure we free it.

dougnazar committed:
Fix various warnings when building with -Wall.

dougnazar committed:
Split my_system() into seperate functions for child & parent.

dougnazar committed:
Rewrite my_system_*() IO loops to handle large buffers.

In my_system_parent(), don't wait for child to exit before reading buffer.
The child may fill the pipe buffer and start erroring, so do both operations
concurrently.

In my_system_child(), similarly check that data/space is available before
we read/write to handle asymetric data rates.

dougnazar committed:
Fix debug display of acl IP addresses

inet_ntoa uses a static buffer which causes duplicate IP addresses to be printed.

dougnazar committed:
Show correct address family when setting up listen addresses.

dougnazar committed:
Only make requested binary

dougnazar committed:
Fix a few pedantic warnings, mostly prototype related.

dougnazar committed:
Fix use of uninitialized variable.

Spotted via valgrind.

dougnazar committed:
Clear any existing ACLs before parsing new config.

Multiple allowed_hosts config lines or every SIGHUP would cause duplicates.
Free previous allowed_hosts to prevent leaking.
Fix dumping of ACLs to properly display AF_INET6 addresses.

dougnazar committed:
Fix various leaks detected by valgrind.

Mostly config options that strdup() their value, causing leaks if
there are multiple config lines or after restarting with SIGHUP.

The last was the fd_set for incoming connections wouldn't be freed
on restart. Also optimize the allocation of fdset and re-use if possible.

dougnazar committed:
Fix minor leak in clean_environ()

my_strpos() modifies the string pointer causing us to free
an incorrect pointer.

rpv-tomsk commented on src/nrpe.c:
```diff
@@ -980,13 +980,15 @@ int read_config_file(char *filename)
+			free(nrpe_user);
 			nrpe_user = strdup(varvalue);
 
-		else if (!strcmp(varname, "nrpe_group"))
```
As a maintainer of other OSS project, personally I dislike such changes, which made without explicit needs. Merging such changes breaks other pull requests and [local] patches of project users, so forces all these people to do extra work just because you like other style of formatting. Also, these changes spray attention of reviewers, while in overall your proposals deserve attention. 

Thanks for your work, I found it useful for me.

dougnazar replied:
I'm confused as to the issue here.

This specific change is required to free any previously `strdup()` config value if it encounters another `nrpe_user` option in the config file or during configuration reload, otherwise we'll leak memory.

If you're referring to the brackets, those are also required as we're switching from a single statement to a compound statement block after the `if` clause.

rpv-tomsk replied:
You completely correct, and I'm completely wrong.
Looking into code yesterday, I saw only brakets and nothing more. Sorry.

//I thought that was only cosmetic change, such as moving bracket to same line with 'if'.

dougnazar committed:
asprintf fixes

dougnazar committed:
Don't include engine.h if not needed.

OpenSSL engines have been deprecated since version 3.0 and at
least Fedora 41 is no longer including the header by default.

dougnazar committed:
Misc minor cleanups.

Fix various [-Wunused-parameter], [-Wsign-compare], [-Wshift-negative-value], etc.
warnings that exist with recent compilers.

dylan-at-nagios commented:
Do you think you could resolve the conflicts for this PR?

dougnazar commented:
> Do you think you could resolve the conflicts for this PR?

@dylan-at-nagios, this should be good to go. There were a few issues that I'd found while working on #291 that I've back ported.

I've some other work I'm interested in doing (ie. with `ssl_recvall`/`ssl_sendall` the ssl/nossl paths are almost identical and should be consolidated) but I'll wait for these patches to be looked at.

ekapsner-ne committed:
Merge branch 'master' into fixes

ekapsner-ne reviewed (approved):
Looks good to me
